src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,1905 @@
+  const Register save_ndl_len = r12;
+
+  __ push(rbp);
+  __ subptr(rsp, STACK_SPACE);
+
+  // Assume failure
+  __ movq(rbp, -1);
+
+  if (isReallyUL) {
+    // Branch out if doing UL
+    __ jmp(L_wcharBegin);
+  }
+
+  if (!isReallyUL && isUU) {  // Adjust sizes of hs and needle
+    // UU passes lengths in terms of chars - convert to bytes
+    __ shlq(needle_len, 1);
+    __ shlq(haystack_len, 1);
+  }
+
+  // UL processing comes here after expanding needle
+  __ bind(L_continue);
+  // nMinusK (haystack length in bytes minus needle length in bytes) is used several
+  // places to determine whether a compare will read past the end of the haystack.
+  __ movq(nMinusK, haystack_len);
+  __ subq(nMinusK, needle_len);
+
+  // Set up expected registers
+  __ movq(save_ndl_len, needle_len);
+  __ movq(r14, needle);
+  __ movq(rbx, haystack);
+
+  // Always need needle broadcast to ymm registers (XMM_BYTE_0 and XMM_BYTE_K)
+  broadcast_first_and_last_needle(needle, needle_len, rax, ae, _masm);
+
+  // Do "big switch" if haystack size > 32
+  __ cmpq(haystack_len, 0x20);
+  __ ja_b(L_bigSwitchTop);
+
+  // Copy the small (< 32 byte) haystack to the stack.  Allows for vector reads without page fault
+  // Only done for small haystacks
+  //
+  // NOTE: This code assumes that the haystack points to a java array type AND there are
+  //       at least 16 bytes of header preceeding the haystack pointer.
+  //
+  // This means that we're copying up to 15 bytes of the header onto the stack along
+  // with the haystack bytes.  After the copy completes, we adjust the haystack pointer
+  // to the valid haystack bytes on the stack.
+  {
+    Label L_moreThan16, L_adjustHaystack;
+
+    const Register index = rax;
+    const Register haystack = rbx;
+
+    // Only a single vector load/store of either 16 or 32 bytes
+    __ cmpq(haystack_len, 0x10);
+    __ ja_b(L_moreThan16);
+
+    __ movq(index, COPIED_HAYSTACK_STACK_OFFSET + 0x10);
+    __ movdqu(XMM_TMP1, Address(haystack, haystack_len, Address::times_1, -0x10));
+    __ movdqu(Address(rsp, COPIED_HAYSTACK_STACK_OFFSET), XMM_TMP1);
+    __ jmpb(L_adjustHaystack);
+
+    __ bind(L_moreThan16);
+    __ movq(index, COPIED_HAYSTACK_STACK_OFFSET + 0x20);
+    __ vmovdqu(XMM_TMP1, Address(haystack, haystack_len, Address::times_1, -0x20));
+    __ vmovdqu(Address(rsp, COPIED_HAYSTACK_STACK_OFFSET), XMM_TMP1);
+
+    // Point the haystack at the correct location of the first byte of the "real" haystack on the stack
+    __ bind(L_adjustHaystack);
+    __ subq(index, haystack_len);
+    __ leaq(haystack, Address(rsp, index, Address::times_1));
+  }
+
+  // Dispatch to handlers for small needle and small haystack
+  // Note that needle sizes of 1-6 have been handled in highly_optimized_short_cases,
+  // so the dispatch only has valid entries for 7-10.
+  __ leaq(r13, Address(save_ndl_len, -1));
+  __ cmpq(r13, NUMBER_OF_CASES - 1);
+  __ ja(L_smallCaseDefault);
+  __ mov64(r15, (int64_t)small_jump_table);
```

Comment:
It may, but I believe the movq is shorter (although maybe not to r15).  I'll do some experimentation.

Reply:
the RIP-relative lea should have a shorter encoding. I think something like `lea(r15, ExternalAddress(small_jump_table))` should produce it (untested)

Reply:
Just did the experiment and it turns out that `mov64(r15, (int64_t)small_jump_table)` and `lea(r15, ExternalAddress(small_jump_table))` produce exactly the same code:

`0x00007fffe463d68b:  49 bf a0 d5 63 e4 ff 7f 00 00   movabs r15,0x7fffe463d5a0`

The code in `MacroAssembler` for `lea` calls `mov_literal64` with no check for whether it can be ip-relative.

I tried doing it myself via `leaq(r15, Address(rip, (int64_t)small_jump_table - (int64_t)(__ pc())))` but there is no definition in `register_x86.hpp` for register `rip`.  So I'm not sure exactly how to produce RIP-relative addressing.

Reply:
Thanks for checking. Well I know that the `MacroAssembler::movdqu(XMMRegister dst, AddressLiteral src, Register rscratch)` method actually generates rip-relative addresses. Maybe we could copy some of that code.

Reply:
Use `lea` and `InternalAddress()` for referencing jump tables since the addresses are in the same code section.

Reply:
Right. Our assembler implements rip-relative addressing for some instructions, but apparently lea isn't one of them.
I'll experiment with it.